docker/Dockerfile

```diff
@@ -4,10 +4,49 @@ ARG PYUWSGI_VER
 ################################ Overview
 # This builds the following hierarchy of images:
 #
-# python:${PYTHON_VER}-slim --> (base) --> (dependencies) --> (dependencies-dev-platform-ARCH) --> (dependencies-dev) -->  dev
-#                                 |              .                                                        |
-#                                 V              .                                                        V
-#                               final <...........                                                    final-dev
+# python:${PYTHON_VER}-slim
+#            |
+#            V
+#         (base)
+#            |----------------------------------------------------------
+#            V                                                          |
+#     (dependencies)                                                    |
+#            |                                                          V
+#            |----------------> (build-nautobot) . . . . . . . . . >  final
+#            V                        /\               .
+# (dependencies-dev-python)            .               .
+#            |                         .               .
+#            |-----------------> (build-docs)          .
+#            V                                         .
+# (dependencies-dev-platform-ARCH)                     .
+#            |                                         .
+#            V                                         .
+#    (dependencies-dev) < . . . . . . . . . . . . . . .
+#       |          |
+#       V          V
+#      dev     final-dev
+#
+#  The design philosphy for this image hierarchy is:
+#  
+#  - pyproject.toml is the source of truth for dependencies
+#  - Non-intermediate image targets (final, final-dev, dev) will not contain files they don't need
+#      - EX: No dev dependencies in final image
+#  - When changing source code only (no dependency changes):
+#      - Dependencies will NOT need to be reinstalled (cached, time intensive)
+#      - Docs will NOT need to be rebuilt
```

Comment:
One thought towards the future - at some point we may want to add Python API documentation to our docs (using something like Sphinx `autodoc` or whatever mkdocs offers as an equivalent). In that eventuality, source code changes *will* require rebuilding the docs. Not sure how that will necessitate changes to the set of build stages here but maybe worth thinking about?

Reply:
So when we do that we'll likely need to copy the source code into the docs generation step and then that would invalidate the cache in that layer.